proposals/p3802.md

```diff
@@ -0,0 +1,1183 @@
+`Box(T)`:
+
+```carbon
+class Box(T:! type) {
+  var ptr: T*;
+  extend api T;
+}
+```
+
+This means that lookup into `Box(T)` also looks in the type `T` for members. In
+this way, `b.Search` will find `b.(String.Search)`.
+
+For the second ingredient, the binding interfaces from
+[proposal #3720: "Binding operators"](https://github.com/carbon-language/carbon-lang/pull/3720)
+already provide everything needed, we just need to make a parameterized
+implementation of them:
+
+```carbon
+impl forall [T:! type, U:! ValueBind(T)]
+    U as ValueBind(Box(T)) where .Result = U.Result {
+  fn Op[self: Self](x: Box(T)) -> Result {
+    // Uses `ValueBind(T).Op` based on type of `U`.
+    return self.Op(*x.ptr);
+
+    // NOT `return x.ptr->(self)` since that attempts
+    // to do reference binding not value binding.
+  }
+}
+
+impl forall [T:! type, U:! RefBind(T)]
+    U as RefBind(Box(T)) where .Result = U.Result {
+  fn Op[self: Self](p: Box(T)*) -> Result* {
+    return self.Op(p->ptr);
+    // Or equivalently:
+    // return p->ptr->(self);
+  }
+}
+```
+
+A few observations:
+
+-   These declarations use `Box` to satisfy the orphan rule, and so this
+    approach only works with `Box(T)` values, not types that can implicitly
+    convert to `Box(T)`.
+-   The implementation of the `Op` method is where we follow the pointer member
+    `ptr` of `Box(T)` to get a `T` value that is compatible with the member
+    being bound. This resolves the type mismatch that is introduced by allowing
+    name resolution to find another type's members.
+-   We have to be a little careful in the implementation of `ValueBind(Box(T))`
+    to still use value binding even when we get a reference expression from
+    dereferencing the pointer `ptr`.
+
+With these two ingredients, `b.Search(32)` is equivalent to
+`b.(String.Search)(32)` due to `extend api`, which then uses the custom binding
+implementation to get something equivalent to `b.ptr->(String.Search)(32)`,
+which is the same as `b.ptr->Search(32)` since `b.ptr` has type `String*`.
+
+### Use case: Reference library type
+
+The `Box` type that was used as
+[an example of member forwarding](#use-case-member-forwarding) preserves the
+expression category -- you can only call mutating operations on a `Box`
+reference expression, not on a `Box` value expression. For a library reference
+type that would simulate the C++-built-in reference type, we instead want to
+permit value binding for anything where `T` permits reference binding. This can
+also be done using `extend api` and a different set of binding implementations:
+
+```carbon
+class Ref(T:! type) {
+  var ptr: T*;
+  extend api T;
+}
+impl forall [T:! type, U:! RefBind(T)]
+    U as ValueBind(Ref(T)) where .Result = Ref(U.Result) {
+  fn Op[self: Self](p: Ref(T)) -> Ref(Result) {
+    return {.ptr = self.Op(p->ptr)};
+  }
+}
+// (Plus impls for the other three combinations of ValueBind and RefBind.)
```

Comment:
Exploring these myself:

```carbon
// <Ref(T) value>.<ref bindable> produces a Ref(R) value expression.
impl forall [T:! type, U:! RefBind(T)]
    U as ValueBind(Ref(T)) where .Result = Ref(U.Result) {
  fn Op[self: Self](p: Ref(T)) -> Result {
    return {.ptr = self.Op(p->ptr)};
  }
}

// <Ref(T) ref expr>.<ref bindable> produces a reference expression.
// Reference collapse: no added Ref(...) here.
impl forall [T:! type, U:! RefBind(T)]
    U as RefBind(Ref(T)) where .Result = U.Result {
  fn Op[self: Self](p: Ref(T)*) -> Result* {
    return self.Op(p->ptr);
  }
}

// <Ref(T) value>.<value bindable> produces a value expression.
// A value binding is performed to turn the Ref(T) into a T.
impl forall [T:! type, U:! ValueBind(T)]
    U as ValueBind(Ref(T)) where .Result = U.Result {
  fn Op[self: Self](p: Ref(T)) -> Result {
    return self.Op(*p.ptr);
  }
}
```

I think implementing `U as RefBind(Ref(T))` in terms of `U as ValueBind(T)` is not possible. We can't produce a pointer as a result of the binding.

There's also overlap between the first and third `impl`s. And the first and second violate the new coherence rule in #3720:  given `ref: Ref({.n: i32})`, `ref.n` will either be a value expression of type `Ref(i32)` or a reference expression of type `i32` depending on the expression category of `ref`.

I think maybe the right thing to do is to drop the second `impl` entirely, and have only two `impl`s, both for `ValueBind`, with a `match_first`:

```carbon
match_first {

// <Ref(T) value>.<ref bindable> produces a value expression of type `Ref(R)`.
impl forall [T:! type, U:! RefBind(T)]
    U as ValueBind(Ref(T)) where .Result = Ref(U.Result) {
  fn Op[self: Self](p: Ref(T)) -> Result {
    return {.ptr = self.Op(p->ptr)};
  }
}

// <Ref(T) value>.<value bindable> produces a value expression of type `R`.
// A value binding is performed to turn the Ref(T) into a T.
impl forall [T:! type, U:! ValueBind(T)]
    U as ValueBind(Ref(T)) where .Result = U.Result {
  fn Op[self: Self](p: Ref(T)) -> Result {
    return self.Op(*p.ptr);
  }
}

}
```

For this to really work, I think we'd need binding to fall back from `RefBind` to `ValueBind` if reference binding isn't possible, as discussed in #3720. The above pair of `impl`s would then be implementing the same fallback semantics for `Ref(T)`.

Reply:
I don't think this works. #3720 only ever produces a reference expression (by automatically dereferencing the returned pointer) when the expression on the left side of the `.` is a reference expression. I think this makes `Ref(T)` impossible, and I should probably delete this section.

Reply:
Sorry, I had not understood what you were proposing (that you can't get a reference expression, but can get a `Ref(T)` value that acts like one). I've made these changes, added the additional support for calls on `Ref(T)` values, and moved this section to future work, [as discussed in open discussion on 05-16](https://docs.google.com/document/d/1s3mMCupmuSpWOFJGnvjoElcBIe2aoaysTIdyczvKX84/edit?resourcekey=0-G095Wc3sR6pW1hLJbGgE0g&tab=t.0#heading=h.3qlpp5e56u46).